lib/motors/controller.cpp

```diff
@@ -1,9 +1,105 @@
+               "Missing required field 'config_messages' in can message file at %s",
+               message_file_path.c_str());
+    return std::nullopt;
+  }
+  if (!document.HasMember("messages")) {
+    logger.log(core::LogLevel::kFatal,
+               "Missing required field 'messages' in can message file at %s",
+               message_file_path.c_str());
+    return std::nullopt;
+  }
+  const auto configuration_messages = document["config_messages"].GetArray();
+  std::vector<core::CanFrame> controller_configuration_messages;
+  for (auto &message : configuration_messages) {
+    const auto new_message = Controller::parseJsonCanFrame(message.GetObject(), logger);
+    if (!new_message) {
+      logger.log(core::LogLevel::kFatal,
+                 "Invalid CAN configuration frame in JSON message file at path %s",
+                 message_file_path.c_str());
+      return std::nullopt;
+    }
+    controller_configuration_messages.push_back(*new_message);
+  }
+  std::unordered_map<std::string, core::CanFrame> controller_messages;
+  const auto messages = document["messages"].GetObject();
+  for (auto &message : messages) {
+    const auto new_message = Controller::parseJsonCanFrame(message.value.GetObject(), logger);
+    if (!new_message) {
+      logger.log(core::LogLevel::kFatal,
+                 "Invalid CAN frame in JSON message file at path %s",
+                 message_file_path.c_str());
+      return std::nullopt;
+    }
+    controller_messages.emplace(message.name.GetString(), *new_message);
+  }
+  return Controller(logger, controller_messages, controller_configuration_messages);
+}
+
+Controller::Controller(core::ILogger &logger,
+                       const std::unordered_map<std::string, core::CanFrame> messages,
+                       const std::vector<core::CanFrame> configuration_messages)
+    : logger_(logger),
+      configuration_messages_(configuration_messages),
+      messages_(messages)
+{
+}
+
+std::optional<core::CanFrame> Controller::parseJsonCanFrame(
+  rapidjson::GenericObject<false, rapidjson::Value> message, core::ILogger &logger)
+{
+  if (!message.HasMember("id")) {
+    logger.log(core::LogLevel::kFatal,
+               "Missing required field 'id' in message in CAN message file");
+    return std::nullopt;
+  }
+  if (!message.HasMember("index")) {
+    logger.log(core::LogLevel::kFatal,
+               "Missing required field 'index' in message in CAN message file");
+    return std::nullopt;
+  }
+  if (!message.HasMember("subindex")) {
+    logger.log(core::LogLevel::kFatal,
+               "Missing required field 'subindex' in message in CAN message file");
+    return std::nullopt;
+  }
+  if (!message.HasMember("data")) {
+    logger.log(core::LogLevel::kFatal,
+               "Missing required field 'data' in message in CAN message file");
+    return std::nullopt;
+  }
+  core::CanFrame new_message;
+  new_message.can_id  = message["id"].GetInt();
+  new_message.can_dlc = 8;
```

Comment:
Why 8? Is this an obvious thing or does it need to be named and defined somewhere?

Reply:
this is the number of bytes in the message, is always 8 as that's the non-extended CAN standard

Reply:
If it's a constant number that has a meaning then should we either give it a name and store it somewhere or put a comment explaining where it comes from so that future hyped members don't have to deal with a load of "magic numbers" which they don't know the purpose of?